File: android/app/src/main/java/com/casttv/androidtv/crypto/Encryption.kt
package com.casttv.androidtv.crypto

import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-GCM encryption matching the Swift CryptoKit implementation.
 *
 * Wire format: nonce (12 bytes) + ciphertext + GCM tag (16 bytes)
 * This matches AES.GCM.SealedBox.combined in Swift.
 */
object Encryption {
    private const val NONCE_LEN = 12
    private const val TAG_LEN_BITS = 128 // 16 bytes

    fun generateKey(): ByteArray {
        val kg = KeyGenerator.getInstance("AES")
        kg.init(256)
        return kg.generateKey().encoded
    }

    /** Encrypt plaintext. Returns nonce (12) + ciphertext+tag. */
    fun encrypt(
        plaintext: ByteArray,
        key: ByteArray,
    ): ByteArray {
        val nonce = ByteArray(NONCE_LEN).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN_BITS, nonce))
        val ciphertextAndTag = cipher.doFinal(plaintext)
        return nonce + ciphertextAndTag
    }

    /** Decrypt combined (nonce + ciphertext+tag) produced by encrypt(). */
    fun decrypt(
        combined: ByteArray,
        key: ByteArray,
    ): ByteArray {
        val nonce = combined.sliceArray(0 until NONCE_LEN)
        val ciphertextAndTag = combined.sliceArray(NONCE_LEN until combined.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN_BITS, nonce))
        return cipher.doFinal(ciphertextAndTag)
    }

    /** Encode key as base64url (no padding) — matches Swift's base64URLEncoded(). */
    fun keyToBase64Url(key: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(key)

    /** Decode base64url key. */
    fun keyFromBase64Url(encoded: String): ByteArray = Base64.getUrlDecoder().decode(encoded)

    /** Encode encrypted payload as standard base64 for WebSocket transport. */
    fun encryptedToBase64(data: ByteArray): String = Base64.getEncoder().encodeToString(data)

    /** Decode base64 WebSocket payload. */
    fun base64ToBytes(encoded: String): ByteArray = Base64.getDecoder().decode(encoded)
}

File: apple/CastTVShared/Sources/CastTVShared/Crypto/Encryption.swift
import Foundation
import CryptoKit

/// AES-GCM encryption/decryption for end-to-end encrypted messaging.
public enum Encryption {

    /// Generate a new AES-256 key.
    public static func generateKey() -> SymmetricKey {
        SymmetricKey(size: .bits256)
    }

    /// Export key as raw bytes.
    public static func exportKey(_ key: SymmetricKey) -> Data {
        key.withUnsafeBytes { Data($0) }
    }

    /// Import key from raw bytes.
    public static func importKey(_ data: Data) -> SymmetricKey {
        SymmetricKey(data: data)
    }

    /// Encrypt plaintext data with AES-GCM.
    /// Returns: nonce (12 bytes) + ciphertext + tag (16 bytes).
    public static func encrypt(_ plaintext: Data, key: SymmetricKey) throws -> Data {
        let sealedBox = try AES.GCM.seal(plaintext, using: key)
        guard let combined = sealedBox.combined else {
            throw EncryptionError.sealFailed
        }
        return combined
    }

    /// Decrypt data encrypted with `encrypt`.
    public static func decrypt(_ combined: Data, key: SymmetricKey) throws -> Data {
        let sealedBox = try AES.GCM.SealedBox(combined: combined)
        return try AES.GCM.open(sealedBox, using: key)
    }

    public enum EncryptionError: Error {
        case sealFailed
    }
}
